### DISCUSSION TITLE
Is it possible for assets to load inputs from ops or vice versa?

### DISCUSSION THREAD

**Discussion by clairelin135** [dagster-io/dagster#14604] [Q&A] [answered]
Is it possible to have an op upstream of an asset? Or have an op downstream of an asset?

**Comment by clairelin135** [comment, marked as answer]
At the moment, this is not possible. Assets can only load inputs and pass outputs to other assets.

If you wanted to reuse an op or a graph as an asset dependency, you can consider creating an op-backed asset or a graph-backed asset.

```
@op 
def my_op(x):
   ...

my_asset = AssetsDefinition.from_op(my_op)
```

For a graph-backed asset:
```
@graph_asset
def my_asset(x):
    return my_op(x)
```

**Reply by marcilj** [reply]
Hey @clairelin135, I've tried something really simple related to using Ops as inputs for Assets and I always receive the following error:
`Error loading repository location my_dagster_project:dagster._core.errors.DagsterInvalidDefinitionError: In @graph trying_graph, received invalid type <class 'dagster._core.definitions.assets.AssetsDefinition'> for input "test_op" (at position 0) in op invocation "test_asset_and_ops". Must pass the output from previous node invocations or inputs to the composition function as inputs when invoking nodes during composition.`

Here's a code example:

```python
@op
def get_test_op(context):
    return 1


get_test_asset = AssetsDefinition.from_op(get_test_op)


@asset
def test_asset_and_ops(test_op):
    return 2 * test_op


@graph
def trying_graph():
    return test_asset_and_ops(get_test_asset)
```

Here's another that return this error instead:
` Error loading repository location my_dagster_project:dagster._core.errors.DagsterInvalidDefinitionError: Input asset '["get_test_asset"]' for asset '["test_asset_and_ops"]' is not produced by any of the provided asset ops and is not one of the provided sources`
```python
@op
def get_test_op(context):
    return 1


get_test_asset = AssetsDefinition.from_op(get_test_op)


@asset
def test_asset_and_ops(get_test_asset):
    return 2 * get_test_asset
```

Can you help me find what I'm missing here?

Thanks.

**Reply by OwenKephart** [reply]
hi @marcilj -- your first example does not work because you're trying to make a graph out of an asset, whereas dagster only allows you to make an asset out of a graph. You'll want to structure your code as shown here: https://docs.dagster.io/concepts/assets/graph-backed-assets#defining-graph-backed-assets. Note that the graph is constructed purely out of ops, and then an asset is constructed from the graph.

For your second example, `get_test_asset` is just the variable name, but the actual asset key for `get_test_asset` is (by default) the name of the originally-decorated function. So your asset key is `get_test_op` in this case. This should work if you instead do:

```python
@op
def my_asset_name(context):
    return 1


my_asset = AssetsDefinition.from_op(my_asset_name)


@asset
def test_asset_and_ops(my_asset_name):
    return 2 * my_asset_name
```
